rfcs/20240531-coo-sparse-encoding/README.md

```diff
@@ -0,0 +1,173 @@
+/// The created memory descriptor will describe a memory object that
+/// contains 3 buffers. The buffers have the following meaning and
+/// assigned numbers (index):
+///  - 0: values
+///  - 1: dim0_indices
+///  - 2: dim1_indices
+///
+/// @param memory_desc Output memory descriptor.
+/// @param ndims Number of dimensions.
+/// @param dims Array of dimensions.
+/// @param data_type Elements data type.
+/// @param nnz Number of non-zero entries.
+/// @param indices_dt Data type of indices.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_memory_desc_create_with_coo_encoding(
+    dnnl_memory_desc_t *memory_desc, int ndims, const dnnl_dims_t dims,
+    dnnl_data_type_t data_type, dnnl_dim_t nnz, dnnl_data_type_t indices_dt);
+```
+
+#### C++ API:
+```cpp
+struct memory {
+  struct desc {
+    /// Function for creating a memory descriptor for COO sparse encodings.
+    ///
+    /// The created memory descriptor will describe a memory object that
+    /// contains 3 buffers. The buffers have the following meaning and
+    /// assigned numbers (index):
+    ///  - 0: values
+    ///  - 1: dim0_indices
+    ///  - 2: dim1_indices
+    ///
+    /// @param adims Tensor dimensions.
+    /// @param adata_type Data precision/type.
+    /// @param nnz Number of non-zero entries.
+    /// @param index_dt Data type of indices.
+    /// @param allow_empty A flag signifying whether construction is
+    ///     allowed to fail without throwing an exception. In this case a
+    ///     zero memory descriptor will be constructed. This flag is
+    ///     optional and defaults to false.
+    static desc coo(const dims &adims, data_type adata_type, dim nnz,
+                    data_type index_dt, bool allow_empty = false);
+  }
+}
+```
+
+#### Usage:
+```cpp
+const int nnz = 12;
+auto md_coo = memory::desc::coo({128, 64}, f32, nnz, s32);
+```
+
+The following example shows how size and data type can be queried for the COO memory descriptor for the different buffers. Here, the [`dnnl_memory_desc_get_size_v2()`](https://oneapi-src.github.io/oneDNN/group_dnnl_api_memory.html#doxid-group-dnnl-api-memory-1gad8ada49d1107442436109ec1de73f370) function works for sorted COO in the same manner as for CSR with the exception of buffex index [2]. For CSR, this buffer contains the pointers to the compressed row indices of the sparse data whereas for the COO format, the stored row indices are uncompressed. Hence, querying the size for index [2] returns the same value as index [1]. The same is true for querying the data type with `dnnl_memory_desc_query_v2()` as the data type is shared by the two indices.
+
+```cpp
+const size_t values_size = dnnl_memory_desc_get_size_v2(coo_md, 0),;
+const size_t indices_size = dnnl_memory_desc_get_size_v2(coo_md, 1);
+char *values = malloc(values_size);
+char *dim0_indices = malloc(indices_size);
+char *dim1_indices = malloc(indices_size);
+```
+
+#### Memory Creation Example:
```

Comment:
Can you please also add an example for the C++ API?

Reply:
Fixed. Thanks!